### PR TITLE
[Testing] Fix types in operations for dense pauli strings

### LINKED ISSUE #4835: Operations on `MutableDensePauliString` return a `DensePauliString`
**Description of the issue**
Doing operations like `__pow__`, `__mul__`, `__rmul__`, `__abs__`, `tensor_product` etc. on a `MutableDensePauliString` return a `DensePauliString` because these methods are inherited from `BaseDensePauliString` which returns `DensePauliString(...)` instead of `type(self)(...)` in these methods. 

https://github.com/quantumlib/Cirq/blob/master/cirq-core/cirq/ops/dense_pauli_string.py#L206-L208

**How to reproduce the issue**

```python
>>> import cirq
>>> m = cirq.MutableDensePauliString("Z")
>>> type(m) == type(-m)
False
```

**Cirq version**
0.14.0.dev

### PR DESCRIPTION
**Pull request by alexbouayad**
Fix #4835.

Fix operations like `__pow__`, `__mul__`, ` __rmul__`, ` __abs__`, `tensor_product`, etc, on abstract class `BaseDensePauliString` so that the operations return compatible types when applied on the two concrete subclasses `DensePauliString` and `MutableDensePauliString`.

Also add a few type annotations, and modify a few tests.

One noticeable implied change: multiplying a MutableDensePauliString object with a DensePauliString object (and vise versa) is not allowed anymore.

### CODE REVIEW

File: cirq-core/cirq/ops/dense_pauli_string.py [hunk 1]
```diff
@@ -64,6 +60,9 @@ class BaseDensePauliString(raw_types.Gate, metaclass=abc.ABCMeta):
     Y_VAL = 2
     Z_VAL = 3
 
+    pauli_mask: np.ndarray
+    coefficient: Union[sympy.Basic, complex]
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/dense_pauli_string.py:64]: Why do we need to declare class variables here?

File: cirq-core/cirq/ops/dense_pauli_string.py [hunk 2]
```diff
@@ -52,7 +48,7 @@
     pauli_gates.Z,
 ]
 
-TCls = TypeVar('TCls', bound='BaseDensePauliString')
+T = TypeVar('T', bound='BaseDensePauliString')
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/dense_pauli_string.py:51]: Can we revert the name `T` to original `TCls` to minimize unrelated changes? Both the type variables are identical modulo naming.

File: cirq-core/cirq/ops/dense_pauli_string.py [hunk 3]
```diff
@@ -188,84 +185,89 @@ def __pow__(self, power):
...
+            return cls(
                 pauli_mask=mask,
                 coefficient=self.coefficient * _vectorized_pauli_mul_phase(self.pauli_mask[i], p),
             )
 
         return NotImplemented
 
-    def __rmul__(self, other):
-        if isinstance(other, (sympy.Basic, numbers.Number)):
+    def __rmul__(self: T, other: Any) -> T:  # type: ignore
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/dense_pauli_string.py:254]: Why the type ignore? (here and elsewhere). Should we revert to not having type annotations in these cases instead of explicitly ignoring them?

File: cirq-core/cirq/ops/dense_pauli_string.py [hunk 4]
```diff
@@ -274,16 +276,24 @@ def tensor_product(self, other: 'BaseDensePauliString') -> 'DensePauliString':
...
+            )
 
-    def __abs__(self):
-        return DensePauliString(coefficient=abs(self.coefficient), pauli_mask=self.pauli_mask)
+        raise TypeError(f"Tensoring allowed only with objects of type {type(self)}.")
 
-    def on(self, *qubits) -> 'cirq.PauliString':
+    def __abs__(self: T) -> T:
+        cls = type(self)
+        return cls(coefficient=abs(self.coefficient), pauli_mask=self.pauli_mask)
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/dense_pauli_string.py:294]: nit: Use `type(self)(...)` instead of `cls(...)` here and elsewhere, where `cls` is used only once

**Review by tanujkhattar — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/ops/dense_pauli_string.py (hunk 1 above)
Comment (alexandrebouayad) [reply cirq-core/cirq/ops/dense_pauli_string.py:64]: No need indeed. I don't actually remember adding those two... strange

File: cirq-core/cirq/ops/dense_pauli_string.py (hunk 2 above)
Comment (alexandrebouayad) [reply cirq-core/cirq/ops/dense_pauli_string.py:51]: Sure

File: cirq-core/cirq/ops/dense_pauli_string.py (hunk 3 above)
Comment (alexandrebouayad) [reply cirq-core/cirq/ops/dense_pauli_string.py:254]: The thing here is that `BaseDensePauliString` is a subclass of `raw_types.Gate`. As a result, mypy expects `BaseDensePauliString.__rmul__` to return an object of a subclass of `cirq.LinearCombinationOfGates`. That is not the case, as the operation actually acts on the `coefficient` attribute of the Pauli string. Maybe a natural fix would be for the operations in `raw_types.Gate` to have `Union[Gate, cirq.LinearCombinationOfGates]` as return type instead of `cirq.LinearCombinationOfGates` alone. I'll commit the fix for review.

File: cirq-core/cirq/ops/dense_pauli_string.py (hunk 4 above)
Comment (alexandrebouayad) [reply cirq-core/cirq/ops/dense_pauli_string.py:294]: Ok

File: cirq-core/cirq/ops/dense_pauli_string.py [hunk 5]
```diff
@@ -180,19 +174,20 @@ def _resolve_parameters_(self: TCls, resolver: 'cirq.ParamResolver', recursive:
...
         if isinstance(power, int):
             i_group = [1, +1j, -1, -1j]
             if self.coefficient in i_group:
                 coef = i_group[i_group.index(self.coefficient) * power % 4]
             else:
                 coef = self.coefficient ** power
             if power % 2 == 0:
-                return coef * DensePauliString.eye(len(self))
-            return DensePauliString(coefficient=coef, pauli_mask=self.pauli_mask)
+                return coef * self.eye(len(self))
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/dense_pauli_string.py:185]: This should be `type(self).eye` since `eye` is a class method. We should also have a test to verify this works as expected. 
```suggestion
                return coef * type(self).eye(len(self))
```

File: cirq-core/cirq/ops/dense_pauli_string.py [hunk 6]
```diff
@@ -205,67 +200,69 @@ def __iter__(self) -> Iterator['cirq.Gate']:
...
         return NotImplemented
 
-    def __mul__(self, other):
-        if isinstance(other, BaseDensePauliString):
-            max_len = max(len(self.pauli_mask), len(other.pauli_mask))
-            min_len = min(len(self.pauli_mask), len(other.pauli_mask))
+    def __mul__(self: TCls, other: Union[sympy.Basic, int, float, complex, TCls]) -> TCls:
+        cls = type(self)
+
+        if isinstance(other, cls):
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/dense_pauli_string.py:218]: Does this mean multiplying a `MutableDensePauliString` instance with a `DensePauliString` was allowed before this change but will not be allowed anymore? This sounds like a breaking change and should not be done. 

For binary operations that expect an `other` argument which can be a base dense pauli string, I think we should default to returning a `MutableDensePauliString` if the types of self and other are not same (but both are instances of BaseDensePauliString), else return `type(self)`.  


This should be done for `__mul__`,  `__rmul__`, `tensor_product` etc. Please also add corresponding tests.

File: cirq-core/cirq/ops/dense_pauli_string.py [hunk 7]
```diff
@@ -333,11 +337,11 @@ def __repr__(self) -> str:
             f'coefficient={proper_repr(self.coefficient)})'
         )
 
-    def _commutes_(self, other: Any, atol: float) -> Union[bool, NotImplementedType, None]:
-        if isinstance(other, BaseDensePauliString):
+    def _commutes_(self, other: Any, atol: float) -> bool:
+        if isinstance(other, type(self)):
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/dense_pauli_string.py:341]:
```suggestion
        if isinstance(other, BaseDensePauliString):
```

File: cirq-core/cirq/ops/raw_types.py [hunk 8]
```diff
@@ -275,31 +275,37 @@ def wrap_in_linear_combination(
...
         return self.wrap_in_linear_combination(coefficient=other)
 
-    def __truediv__(self, other: Union[complex, float, int]) -> 'cirq.LinearCombinationOfGates':
+    def __truediv__(
+        self, other: Union[complex, float, int]
+    ) -> Union['Gate', 'cirq.LinearCombinationOfGates']:
         return self.wrap_in_linear_combination(coefficient=1 / other)
 
-    def __pow__(self, power):
+    def __pow__(self, power: Union[float, int]) -> 'Gate':
```
Comment (tanujkhattar) [inline cirq-core/cirq/ops/raw_types.py:308]: `power` can be a symbol as well.
The return value can be `NotImplemented` as well. 

```suggestion
    def __pow__(self, power: Union[float, int, sympy.Basic]) -> Union['Gate', NotImplemented]:
```

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Left another round of comments. Main concern is about potential breaking changes when doing binary operations on dense pauli strings of different type types.

File: cirq-core/cirq/ops/dense_pauli_string.py (hunk 5 above)
Comment (alexandrebouayad) [reply cirq-core/cirq/ops/dense_pauli_string.py:210]: - `self.eye` and `type(self).eye` are semantically equivalent here. Is your suggestion referring to some convention or best practice?
- I'll add the test

**Comment by vtomole** [comment]
Friendly ping @alexandrebouayad: Do you have time to finish this up? I can take it on if not.

**Comment by alexandrebouayad** [comment]
> Friendly ping @alexandrebouayad: Do you have time to finish this up? I can take it on if not.

@vtomole thanks for the reminder, it had slipped my mind. I'll do it in the coming days.

File: cirq-core/cirq/ops/dense_pauli_string.py (hunk 3 above)
Comment (alexandrebouayad) [reply cirq-core/cirq/ops/dense_pauli_string.py]: I rollbacked the previous change, that is, the return type of linear operations  in `raw_types.py` is now back to `cirq.LinearCombinationOfGates` alone. I used `Gate`'s method `wrap_in_linear_combination` to convert the results of linear operations in `dense_pauli_string.py` (I mimicked what's done in `raw_types.py`).

File: cirq-core/cirq/ops/dense_pauli_string.py (hunk 5 above)
Comment (alexandrebouayad) [reply cirq-core/cirq/ops/dense_pauli_string.py]: I changed to `type(self)` as suggested.

File: cirq-core/cirq/ops/dense_pauli_string.py (hunk 6 above)
Comment (alexandrebouayad) [reply cirq-core/cirq/ops/dense_pauli_string.py]: I added the possibility to cross operate on mutable and non-mutable dense Pauli strings.

**Comment by alexandrebouayad** [comment]
I have made the changes as suggested, plus a few other minor ones. I'll add the remaining tests, and I'll ensure that local checks pass by the end of this week hopefully.

**Comment by tanujkhattar** [comment]
@alexandrebouayad Can you please fix the remaining failing tests? Thanks!

**Comment by tanujkhattar** [comment]
@alexandrebouayad Gentle reminder, do you still have time to work on this?

**Comment by vtomole** [comment]
Hey @alexandrebouayad , I'm pretty sure this PR has been superseded by https://github.com/quantumlib/Cirq/pull/5624. We would love your review on that PR.

**Comment by tanujkhattar** [comment]
@vtomole Actually, I did skip over some of the tricky parts of fixing mypy types in my PR https://github.com/quantumlib/Cirq/pull/5624; for example: 

Adding return types annotations to operations like `__neg__`, `__add__` etc. breaks mypy because the `cirq.Gate` class returns `LinearCombinationOfGates` by default for these methods, but in `BaseDensePauliString`, we want to return `type(self)(...)`, which is a (subclass of) `BaseDensePauliString` and not a `LinearCombinationOfGates`. This breaks Liskov Substitution Principle because `BaseDensePauliString`  is not a subclass of `LinearCombinationOfGates`.

Fixing this would require us to modify the default return types on `cirq.Gate`, which will be a big breaking change. We can potentially open a new issue to track this and maybe @alexandrebouayad's PR can tackle the new issue if he's already thought on this (I see the solution he has applied in his PR is to completely remove the return types from `cirq.Gate`'s methods -- I'm not sure if this is the best we can do though; it's probably better to not have return types on a specific implementation  like `BaseDensePauliString` instead of removing return types from `cirq.Gate`)
